Update gradle

lognaturel opened the pull request:
Gradle 4.7 has been out for some time now and it's easier to do incremental updates than to do big updates less frequently.

Updates other dependencies that don't require code changes or a Play Services upgrade.

This also introduces variables for versions that are common between multiple dependencies as suggested by @jd-alexander. Not all versions are handled through variables, only the ones that are for sure always common. If someone disagrees with this decision, let's open a new issue and have a discussion there. 

The introduction of those version variables makes dependency declarations longer so I've standardized on using string notation over map notation for that reason.

lognaturel committed:
Update Gradle plugin to 3.1.2

The plugin now includes a default build tools version so it's recommended not to set an explicit version. androidTestCompile is now deprecated and replaced by androidTestImplementation.

lognaturel committed:
Update dependencies (no code changes, no play services update)

lognaturel committed:
Upgrade gradle to 4.7

lognaturel committed:
Address new lint issues

lognaturel committed:
Force consistent findbugs version

shobhitagarwal1612 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,9 @@
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.5.1-bin.zip
-android.enableD8.desugaring=true
\ No newline at end of file
+android.enableD8.desugaring=true
```
I thought adding this would address the failing circle ci builds, but they are failing again occasionally with the same error. I feel like this fix isn't working and can be removed

lognaturel replied:
I'm not totally sure. It does feel like there have been fewer failures and it doesn't seem like there are downsides so I'd say we can leave it in a little longer.

yanokwa replied:
I did some quick research and I think enabling desugaring is a good thing regardless of the failing builds.

jd-alexander commented on collect_app/build.gradle:
```diff
@@ -176,14 +177,14 @@ allprojects {
+    implementation group: 'com.android.support', name: 'support-v13', version: '27.1.1'
+    implementation group: 'com.android.support', name: 'customtabs', version: '27.1.1'
+    implementation group: 'com.android.support', name: 'design', version: '27.1.1'
+    implementation group: 'com.android.support', name: 'appcompat-v7', version: '27.1.1'
```
I was thinking that variables could be introduced in the gradle file to reduce the amount of work necessary to update dependencies and also to make it easier for conflict resolution. for eg., a `SUPPORT_VERSION` variable could be referenced by all these support related dependencies and the version change just takes place in one place. This is how it's done in a popular MVP project : [variables](https://github.com/MindorksOpenSource/android-mvp-architecture/blob/master/build.gradle). [dependencies](https://github.com/MindorksOpenSource/android-mvp-architecture/blob/master/app/build.gradle)
What do you think?

lognaturel replied:
Great idea, will do!

jd-alexander commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,9 @@
-android.enableD8.desugaring=true
\ No newline at end of file
+android.enableD8.desugaring=true
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.7-all.zip
```
We probably want to leave this as `gradle-4.7-bin.zip` so we don't end up pulling in all the documentation and sources since the binaries are good enough. Android Studio is always changing it. I think it's trying to force us to read gradle docs 🤣

lognaturel replied:
Ahh, yes, thanks. 😊

jd-alexander commented on collect_app/build.gradle:
```diff
@@ -220,7 +221,7 @@ dependencies {
-    implementation group: 'com.jakewharton.timber', name: 'timber', version: '4.6.0'
+    implementation group: 'com.jakewharton.timber', name: 'timber', version: '4.7.0'
     implementation group: 'com.google.zxing', name: 'core', version: '3.3.0'
     implementation group: 'com.journeyapps', name: 'zxing-android-embedded', version: '3.5.0'
```
In terms of notation, what's the reason for the usage of both the string and map notation for specifying dependencies? Just curious because when I am adding a dependency I am wondering which one I should use 😄

lognaturel replied:
We had standardized on the map notation but it looks like some PRs got in at some point using the string notation. Would be good to get everything in line.

lognaturel committed:
Use variables for versions used more than once

Only addresses versions that are absolutely known to be in sync. For example, it's not clear whether rxrelay and rxbinding will always have the same version.

lognaturel committed:
Standardize on string notation and quotes

Introducing version variables leads to longer declarations making string notation more attractive.

yanokwa commented on collect_app/src/debug/AndroidManifest.xml:
```diff
@@ -6,7 +6,7 @@
 
     <!-- Allows changing locales -->
-    <uses-permission android:name="android.permission.CHANGE_CONFIGURATION" />
+    <uses-permission android:name="android.permission.CHANGE_CONFIGURATION" tools:ignore="ProtectedPermissions" />
```
Just for others to know what this is for. CHANGE_CONFIGURATION can only be granted to system apps. Since this is a debug build, we are OK to use this permission, but we do have to tell lint to ignore it.

jd-alexander commented:
Just realized this got merged in! I wanted an update to the `Robolectric` dependency and it has several dependencies as well so it could get its own variable whenever possible.

lognaturel commented:
@jd-alexander You can go ahead and do another little PR for that. Just make sure that all tests still pass -- I seem to remember there was an issue with upgrading but my memory might be failing me!

jd-alexander commented:
Okay no prob!
